samples/snippets/csharp/VS_Snippets_CLR/etwtracelarge/cs/program.cs

```diff
@@ -0,0 +1,199 @@
+//<root>
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.Tracing;
+
+namespace Demo
+{
+    //<InterfaceSource>
+    public interface IMyLogging
+    {
+        void Error(int errorCode, string message);
+        void Warning(string message);
+    }
+
+    public sealed class MySource : EventSource, IMyLogging
+    {
+        public static MySource Log = new();
+
+        [Event(1)]
+        public void Error(int errorCode, string message) => WriteEvent(1, errorCode, message);
+
+        [Event(2)]
+        public void Warning(string message) => WriteEvent(2, message);
+    }
+    //</InterfaceSource>
+
+    //<UtilitySource>
+    public abstract class UtilBaseEventSource : EventSource
+    {
+        protected UtilBaseEventSource() 
+            : base()
+        { }
+
+        protected UtilBaseEventSource(bool throwOnEventWriteErrors)
+            : base(throwOnEventWriteErrors)
+        { }
+
+        // helper overload of WriteEvent for optimizing writing an event containing
+        // payload properties that don't align with a provided overload. This prevents
+        // EventSource from using the object[] overload which is expensive.
+        protected unsafe void WriteEvent(int eventId, int arg1, short arg2, long arg3)
+        {
+            if (IsEnabled())
+            {
+                EventSource.EventData* descrs = stackalloc EventSource.EventData[3];
+                descrs[0] = new EventData { DataPointer = (IntPtr)(&arg1), Size = 4 };
```

Comment:
Is this a safe thing to do in a sample? Here you are getting them from args only on primitives, which makes it OK. Do we have good guidance around fields of classes for things like strings?

Reply:
Event methods on `EventSource` must call `WriteEvent` and their parameters must match. An exception will be thrown if the event method's parameter list doesn't match the parameter list of the `WriteEvent` overload used. This code is adding an overload of `WriteEvent` that knows how to quickly write a payload of `int`, `short`, `long` which isn't a pre-written overload.

The `EventData` struct can _only_ contain types that are capable of being represented with `EventSource` metadata:
```
// A Valid type to log to an EventSource include
//   * Primitive data types
//   * IEnumerable<T> of valid types T (this include arrays)  (* New for V4.6)
//   * Explicitly Opted in class or struct with public property Getters over Valid types.  (* New for V4.6)
```

`EventSource` derives its metadata from the parameter list of the event method which _must_ match the call to `WriteEvent`.

In other words, you can only do this optimization with strings and primitive types. You could theoretically write something like this though:

```csharp
public class MyType
{
	public string Foo { get; set; }
	public int Bar { get; set; }
	public long Baz { get; set; }
}

public abstract class UtilSource : EventSource
{
	protected UtilSource() : base () { }
	protected UtilSource(bool throwOnEventWriteErrors) : base(throwOnEventWriteErrors) { }

	protected unsafe void WriteEvent(int eventId, string arg1, int arg2, long arg3)
	{
		if (IsEnabled())
		{
			fixed (char* _arg1 = arg1 ?? string.Empty)
			{
				EventSource.EventData* descrs = stackalloc EventSource.EventData[3];
				descrs[0] = new EventData { DataPointer = (IntPtr)(_arg1), Size = ((arg1?.Length ?? 0) + 1) * 2 };
				descrs[1] = new EventData { DataPointer = (IntPtr)(&arg2), Size = 4 };
				descrs[2] = new EventData { DataPointer = (IntPtr)(&arg3), Size = 8 };
			}
		}
	}
}

public sealed MySource : UtilSource
{
	public static MySource Log = new();

	[NonEvent]
	public void LogThing(MyType thing) => DumpMyType(thing.Foo, thing.Bar, thing.Baz);

	[Event(1)]
	public void DumpMyType(string arg1, int arg2, long arg3) => WriteEvent(1, arg1, arg2, arg3);
}
```

Reply:
Mostly what I meant: should the sample show to save pointers for things that are not safe to assume pinned (strings and other things adhering to the third point of "opted in", including value types that live in the heap like boxed and embedded types). Exactly like the example you have or something like `CounterPayload` since this is an advanced scenario type of thing.

Reply:
You can't really do this optimization with anything reference typed besides strings and you need to have the `fixed` scope surround your call to `WriteEventCore` or you will run into issues.

The "opted in" option is a reference to types decorated with `EventDataAttribute` which only applies to the `Write<T>` API which can't be optimized like this.